### PR TITLE
LF: Tests for Node.isReplayedBy

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This PR advances the state of #6665.

This test would have catch the bug fixed in #6664.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/NodeSpec.scala [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+
+      val rightTestCases =
+        Table(
+          "replayed node",
+          create,
+          fetch,
+          lookup,
+          exercise.copy(targetCoid = anotherContractId),
+          exercise.copy(templateId = AnotherTemplateTyCon),
+          exercise.copy(choiceId = anotherChoideName),
```
Comment (andreaslochbihler-da) [inline daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/NodeSpec.scala:248]:
```suggestion
          exercise.copy(choiceId = anotherChoiceName),
```

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/NodeSpec.scala [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+      val rightTestCases =
+        Table(
+          "replayed node",
+          create,
+          fetch,
+          exercise,
+          lookup.copy(templateId = AnotherTemplateTyCon),
+          lookup.copy(key = lookup.key.copy(key = anotherKey)),
+          lookup.copy(key = lookup.key.copy(key = lookup.key.key.copy(version = v7))),
+          lookup.copy(key = lookup.key.copy(maintainers = aliceAndBob))
```
Comment (andreaslochbihler-da) [inline daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/NodeSpec.scala:300]: Can you add a test case for a different result being returned?

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/NodeSpec.scala [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+      forEvery(leftTestCases)(Node.isReplayedBy(_, create) shouldBe false)
+
+    }
+
+    "accept valid replay of a create node" in {
+
+      val leftTestCases =
+        Table(
+          "replayed node",
+          create,
```
Comment (andreaslochbihler-da) [inline daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/NodeSpec.scala:175]: Do you assume that values have a unique shape? What about a `ValueRecord` where some field names have been omitted, i.e., set to `None`?

Comment (andreaslochbihler-da) [reply daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/NodeSpec.scala:175]: I don't think it's necessary to test this exhaustively here. But IMO it makes sense to include one positive test case for each kind of node to make sure that we don't use `==` on values inside the nodes.

Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/NodeSpec.scala:175]: As far I can tell there is no difference between `` `Value Equal instance` `` and standard equality (expect type safety and contract ID parametrization). 
In other words, "Yes we assume that values have a unique shape". Type constructors or record field names can be omitted only for values coming from the ledger API.

Comment (andreaslochbihler-da) [reply daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/NodeSpec.scala:175]: OK. Makes sense. Then there's no point in testing that here.

**Review by andreaslochbihler-da — APPROVED** [review]

**Comment by cocreature** [comment]
@remyhaemmerle-da what is the state of this? I’ve only briefly skimmed it but given that I’ve pointed out the lack of tests for validation in #8062, I’d be very much in favor of getting this or some variant thereof merged.

**Comment by remyhaemmerle-da** [comment]
@moritzkiefer-da I have first to specify the validation and write a lot of tests. 
I have an issue for that #6665.

**Pull request closed without merging**